src/main/java/seedu/recipe/ui/RecipeForm.java

```diff
@@ -0,0 +1,245 @@
+
+import javafx.fxml.FXML;
+import javafx.scene.Scene;
+import javafx.scene.control.Button;
+import javafx.scene.control.TextField;
+import javafx.scene.input.KeyCode;
+import javafx.scene.layout.FlowPane;
+import javafx.scene.layout.Region;
+import javafx.scene.layout.VBox;
+import javafx.stage.Modality;
+import javafx.stage.Stage;
+import seedu.recipe.model.recipe.Recipe;
+import seedu.recipe.ui.events.EditRecipeEvent;
+
+/**
+ * Represents the form element for users to edit {@code Recipe}s
+ */
+public class RecipeForm extends UiPart<Region> {
+    private static final String FXML = "RecipeForm.fxml";
+
+    @FXML
+    private TextField nameField;
+
+    @FXML
+    private TextField durationField;
+
+    @FXML
+    private TextField portionField;
+
+    @FXML
+    private TextField ingredientsField;
+
+    @FXML
+    private TextField stepsField;
+
+    @FXML
+    private TextField tagsField;
+
+    @FXML
+    private FlowPane tags;
+
+    @FXML
+    private VBox ingredientsBox;
+
+    @FXML
+    private VBox stepsBox;
+
+    @FXML
+    private FlowPane ingredients;
+
+    @FXML
+    private FlowPane steps;
+
+    @FXML
+    private Button saveButton;
+
+    @FXML
+    private Button cancelButton;
+    private int displayedIndex;
+    private Map<String, String> initialValues;
+    private Recipe recipe;
+
+    /**
+     * Creates a new RecipeForm with the given recipe and displayed index.
+     * If the recipe is not null, the form fields are pre-populated with the recipe's data.
+     *
+     * @param recipe         The recipe to edit or null for creating a new recipe.
+     * @param displayedIndex The index of the recipe in the displayed list.
+     */
+    public RecipeForm(Recipe recipe, int displayedIndex) {
+        super(FXML);
+        this.recipe = recipe;
+        this.displayedIndex = displayedIndex;
+        if (recipe != null) {
+            populateFields();
+        }
+        assert saveButton != null;
+        saveButton.setOnAction(event -> saveRecipe());
+        cancelButton.setOnAction(event -> closeForm());
+    }
+
+    /**
+     * Stores the initial values of the form fields in a HashMap.
+     * This is used for comparison when saving the recipe to determine
+     * which fields have been changed.
+     */
+    private void storeInitialValues() {
+        initialValues = new HashMap<>();
+        initialValues.put("name", nameField.getText());
+        initialValues.put("duration", durationField.getText());
+        initialValues.put("portion", portionField.getText());
+        initialValues.put("ingredients", ingredientsBox.getChildren().stream()
+            .map(node -> ((TextField) node).getText())
+            .collect(Collectors.joining(", ")));
+
+        initialValues.put("steps", stepsBox.getChildren().stream()
+            .map(node -> ((TextField) node).getText())
+            .collect(Collectors.joining(", ")));
+
+        initialValues.put("tags", tagsField.getText());
```

Comment:
you may want to abstract out the hashmap keys into class-level final variables. the rationale is since we're going to be using it in multiple places (e.g. in the switch case below), we want to make sure that we use the same set of keys